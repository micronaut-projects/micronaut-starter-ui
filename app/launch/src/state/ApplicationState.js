import { useMemo } from 'react'
import { RecoilRoot } from 'recoil'
import { initializeStateFactory } from './factories/initializeState'

export default function ApplicationState({
  initialData,
  stateInitializer,
  children,
}) {
  const initializeState = useMemo(() => {
    if (typeof stateInitializer === 'function') return stateInitializer
    return initializeStateFactory(initialData)
  }, [initialData, stateInitializer])

  return <RecoilRoot initializeState={initializeState}>{children}</RecoilRoot>
}
