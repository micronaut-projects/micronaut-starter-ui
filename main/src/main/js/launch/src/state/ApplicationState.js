import { RecoilRoot } from 'recoil'
import { initialValueState } from './store'

import {
  initialValueGenerator,
  providedDefaults,
} from './factories/providedDefaults'
import { StarterSDK } from './factories/StarterSDK'

import { ACTIVITY_KEY, isDeepLinkReferral } from '../helpers/Routing'
import { formResets } from './factories/formResets'
import { getStorageValue } from '../hooks/useLocalStorage'

const INITIAL_FORM_DATA_STORAGE_KEY = 'INITIAL_FORM_DATA'

const initialForm = (initialData) => {
  const { javaVersion, lang, build, test, features } = initialData
  const parsed = {
    javaVersion: typeof javaVersion === 'string' ? javaVersion : '', // This is specifically "" to work with the SelectOption component
    lang: typeof lang === 'string' ? lang : '',
    build: typeof build === 'string' ? build : '',
    test: typeof test === 'string' ? test : '',
    features: StarterSDK.reconstructFeatures(features),
    [ACTIVITY_KEY]: initialData[ACTIVITY_KEY],
  }
  return {
    ...parsed,
    ...formResets(initialData),
  }
}

function extractAccessoryData(query = {}) {
  return {
    version:
      query?.version ?? getStorageValue('SELECTED_MN_VERSION', null)?.version,
    debug: query.debug,
  }
}

export default function ApplicationState({ initialData, children }) {
  function initializeState({ set }) {
    const query = providedDefaults()
    const isReferral = isDeepLinkReferral(query)
    const initialValueData = initialValueGenerator(
      INITIAL_FORM_DATA_STORAGE_KEY,
      initialForm(query),
      isReferral
    )
    const accessory = extractAccessoryData(query)
    const init = Object.assign({}, initialValueData, accessory, initialData)

    set(initialValueState, init)
  }

  return <RecoilRoot initializeState={initializeState}>{children}</RecoilRoot>
}
