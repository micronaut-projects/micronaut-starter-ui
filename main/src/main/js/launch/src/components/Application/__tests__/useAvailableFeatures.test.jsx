// Link.react.test.js
import React, { useEffect, useLayoutEffect } from 'react'
import { create, act } from 'react-test-renderer'
import { useRecoilState, useResetRecoilState } from 'recoil'
import ApplicationState from '../../../state/ApplicationState'
import {
  sdkFactoryState,
  sdkState,
  selectedVersionState,
  useApplicationType,
  useAvailableFeatures,
} from '../../../state/store'

const MockTypes = {
  DEFAULT: { features: ['a', 'b', 'c'] },
}
const MockSdk = {
  baseUrl: 'http://pretend.com',
  features: async ({ type }) => {
    const types = MockTypes[type]
    if (!types) {
      throw new Error('Invalid Type')
    }
    return types
  },
}

const useMount = (cb) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useLayoutEffect(() => cb(), [])
}

const TestView = ({ sdk, applicationType, onError }) => {
  const { features, error } = useAvailableFeatures()
  const [, setVersion] = useRecoilState(selectedVersionState)
  const [, setType] = useApplicationType()
  const [, setSdkFactory] = useRecoilState(sdkFactoryState)

  useMount(() => {
    setVersion({ api: sdk.baseUrl })
    setSdkFactory(() => () => sdk)
    setType(applicationType)
  })

  useEffect(() => {
    if (error) onError(error)
  }, [error, onError])

  return <div className="test">{JSON.stringify(features, null, 2)}</div>
}

const TEST_DATA = [
  { initialData: { type: 'DEFAULT' }, hasError: false },
  { initialData: { type: 'BOOM' }, hasError: true },
]

TEST_DATA.forEach(({ initialData, hasError }) => {
  it(`Initial Form Data with "${initialData.type}"`, async () => {
    let error = null
    const onError = (e) => {
      error = e
    }

    let testRenderer
    act(() => {
      testRenderer = create(
        <ApplicationState>
          <TestView
            sdk={MockSdk}
            applicationType={initialData.type}
            onError={onError}
          />
        </ApplicationState>
      )
    })
    await act(() => Promise.resolve())

    expect(testRenderer.toJSON()).toMatchSnapshot()
    if (hasError) {
      expect(error).not.toBeNull()
    } else {
      expect(error).toBeNull()
    }
  })
})
