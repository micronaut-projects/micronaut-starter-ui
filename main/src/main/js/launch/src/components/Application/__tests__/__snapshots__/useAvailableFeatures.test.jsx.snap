// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Form Data with "BOOM" 1`] = `
<div
  className="test"
>
  []
</div>
`;

exports[`Initial Form Data with "DEFAULT" 1`] = `
<div
  className="test"
>
  [
  "a",
  "b",
  "c"
]
</div>
`;
